### DISCUSSION TITLE
PWA Support by default / PageSpeed 100

### DISCUSSION THREAD

**Discussion by hlohaus** [shopware/platform#3160] [RFC / Proposal]
**Effort**: high

### Description: 

It's not possible to add new routes to the pwa extension, because its to complicated and the routes are hard coded.
https://github.com/hlohaus/SwagShopwarePwa/blob/master/src/Pwa/PageLoader/Context/PathResolver.php#L24

A route in shopware 6 are like in Shopware 5 a controller class.
They use a route class or a page loader for loading data. (M)
They render the viewport and define the response type. (V)
They have the route scope and the router route and combined all together. (C)

For a pwa route you need everythings too, but as micro service / in a complete other way.

### Benefits

- Performance, because the page can be send in small parts.
 (The reason why PWA can have PageSpeed 100)
- Less duplicate code (for StoreAPI, PWA and MVC)
- No PWA extension required

### Break strategy

Deprecated old controller and replace them.

**Comment by OliverSkroblin** [comment]
Hey @hlohaus,
thanks for the proposal. I totally agree with you that this place is not extensible at all and is of course suboptimal for third party developers. Since you seem to have dealt with the topic more, it would be cool if you could make one or two suggestions on how this place would be better implemented for you.

For example, the storefront seo-urls are resolved before the `kernel::handle`, but this doesn't seem to be the case with pwa, because `PathResolverInterface` doesn't come from the core. Maybe this would also be an approach

**Comment by hlohaus** [comment]
Hey @OliverSkroblin 
Yes, but my biggest interest is to make pwa to default way.
**PWA First Rule** like Mobile First
A Controller for the main content and a controller for the first view on mobile.
As customer you only see the menu first and then you maybe scroll to the content.
We have also load the content only one second later.

With this trick we have pagespeed 100, without big performance improvements (which not are necessary :) )
